src/patches/mapkey.cc

```diff
@@ -0,0 +1,143 @@
+#include <prime/TMP_InputField.h>
+#include <string>
+#include <string_view>
+
+MapKey::MapKey()
+{
+  this->Key = KeyCode::None;
+}
+
+MapKey* MapKey::Parse(std::string_view key)
+{
+  auto strippedKey = absl::StripTrailingAsciiWhitespace(key);
+  auto lowerKey    = absl::AsciiStrToUpper(strippedKey);
+  auto wantedKeys  = absl::StrSplit(lowerKey, "-", absl::SkipWhitespace());
+
+  auto mapKey = new MapKey();
+  for (auto wantedKey : wantedKeys) {
+    auto modifier = ModifierKey::Parse(wantedKey);
+    if (modifier) {
+      mapKey->Modifiers.emplace_back(modifier);
+      mapKey->Shortcuts.emplace_back(wantedKey);
+    } else {
+      auto parsedKey = Key::Parse(wantedKey);
+
+      if (Key::IsModifier(parsedKey)) {
+        continue;
+      }
+
+      if (parsedKey != KeyCode::None) {
+        mapKey->Key = parsedKey;
+        mapKey->Shortcuts.emplace_back(wantedKey);
+      }
+    }
+  }
+
+  if (mapKey->Key != KeyCode::None) {
+    return mapKey;
+  }
+
+  return nullptr;
+}
+
+bool MapKey::HasGameFunction(GameFunction gameFunction)
+{
+  return MapKey::mappedKeys.contains(gameFunction);
+}
+
+void MapKey::AddMappedKey(GameFunction gameFunction, MapKey* mappedKey)
+{
+  if (MapKey::HasGameFunction(gameFunction)) {
+    return MapKey::SetMappedKey(gameFunction, mappedKey);
+  }
+
+  MapKey::mappedKeys.insert(std::make_pair(gameFunction, mappedKey));
+}
+
+void MapKey::SetMappedKey(GameFunction gameFunction, MapKey* mappedKey)
+{
+  if (!MapKey::HasGameFunction(gameFunction)) {
+    return MapKey::AddMappedKey(gameFunction, mappedKey);
+  }
+
+  MapKey::mappedKeys[gameFunction] = mappedKey;
+}
+
+bool MapKey::IsPressed(GameFunction gameFunction)
+{
+  if (!MapKey::HasGameFunction(gameFunction)) {
+    return false;
+  }
+
+  MapKey* mapKey = MapKey::mappedKeys.at(gameFunction);
+
+  if (Key::Pressed(mapKey->Key)) {
+    return MapKey::HasCorrectModifiers(mapKey);
+  }
+
+  return false;
+}
+
+bool MapKey::IsDown(GameFunction gameFunction)
+{
+  if (!MapKey::HasGameFunction(gameFunction)) {
+    return false;
+  }
+
+  MapKey* mapKey = MapKey::mappedKeys.at(gameFunction);
+
+  if (Key::Down(mapKey->Key)) {
+    return MapKey::HasCorrectModifiers(mapKey);
+  }
+
+  return false;
+}
+
+bool MapKey::HasCorrectModifiers(MapKey* mapKey)
+{
+  auto result = false;
+  if (mapKey->Modifiers.empty()) {
+    result = !Key::IsModified();
+  } else {
+    result = true;
+    for (const auto modifier : mapKey->Modifiers) {
+      if (!modifier->IsPressed()) {
+        result = false;
+        break;
+      }
+    }
+  }
+
+  return result;
+}
+
+std::string MapKey::GetParsedValues()
+{
+  std::string output = "";
+  for (const std::string_view key : this->Shortcuts) {
+    if (output.length()) {
+      output.append("-");
+    }
+    output.append(key);
+  }
+
+  return output;
+}
+
+std::map<GameFunction, MapKey*> MapKey::mappedKeys = {};
+
+std::vector<std::string>  Shortcuts = {};
+std::vector<ModifierKey*> Modifiers = {};
+
+KeyCode Key;
```

Comment:
I am not quite sure what these dare doing here?

Reply:
Not quite sure I follow what you're missing, they are the fields used by the class?  Shortcuts and Modifiers were being used for logging purposes originally, though I guess these could now become private fields.